puzzle-utils/words/dropquote.py

```diff
@@ -0,0 +1,193 @@
+        current_letters.append(column.lower())
+
+    stuck = False
+    dropQuoteSnapshots = []
+    while not stuck:
+        stuck = True
+        dropQuoteSnapshot = dropquote_search(current_grid, current_letters)
+        dropQuoteSnapshots.append(dropQuoteSnapshot)
+
+        for index, words in dropQuoteSnapshot.words.items():
+            if len(words.words) == 1:
+                stuck = False
+                dropquote_apply_word(current_grid, current_letters, index, words.words[0])
+    
+    return dropQuoteSnapshots
+
+def dropquote_validate(grid, letters):
+    row_length = len(grid[0])
+    column_lengths = [0] * row_length
+    for row in range(len(grid)):
+        if row_length != len(grid[row]):
+            raise ValueError(f'Row {row} did not have the same length as the first row.')
+        
+        for column in range(len(grid[row])):
+            if grid[row][column] == ' ':
+                column_lengths[column] += 1
+    
+    for column in range(row_length):
+        if len(letters[column]) != column_lengths[column]:
+            raise ValueError(f'Column {column} has {len(letters[column])} letters but has {column_lengths[column]} blanks.')
+
+def dropquote_search(grid, letters):
+    snapshot = DropQuoteSnapshot()
+    snapshot.grid = grid[:]
+    snapshot.letters = letters[:]
+
+    row = 0
+    column = 0
+    word_index = 0
+    current_word = ''
+    current_letters = []
+    new_word = True
+    while row < len(grid) and column < len(grid[0]):
+        if grid[row][column] == '-' and new_word == False:
+            if current_word.find(' ') != -1:
+                # Overlapped columns could produce a superset of the actual possible words
+                snapshot.words[word_index] = DropQuoteWords(current_letters[:], dropquote_search_words(current_word, current_letters))
+            
+            word_index += 1
+            current_word = ''
+            current_letters = []
+            new_word = True
+        elif grid[row][column] == ' ':
+            current_word += ' '
+            current_letters.append(letters[column])
+            new_word = False
+        else:
+            current_word += grid[row][column]
+            current_letters.append('')
+            new_word = False
+        
+        column += 1
+        if column >= len(grid[row]):
+            row += 1
+            column = 0
+    
+    if current_word.find(' ') != -1:
+        # Overlapped columns could produce a superset of the actual possible words
+        snapshot.words[word_index] = DropQuoteWords(current_letters[:], dropquote_search_words(current_word, current_letters))
+
+    return snapshot
+
+def dropquote_search_words(word, letters):
+    pattern = '^'
+    for index in range(len(word)):
+        if word[index] == ' ':
+            pattern += f'[{letters[index]}]'
+        else:
+            pattern += word[index]
+    pattern += '$'
+    return search(pattern)
+
+def dropquote_apply_word(grid, letters, word_index, word):
```

Comment:
Have you considered putting these functions in a class (called `DropQuoteSolver` perhaps)? Having multiple functions taking in similar arguments (in this case `grid` and `letters`) is usually a sign that they can be put in a class. You can keep `dropquote` function out of the class as the convenience method for the users